cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -147,103 +155,80 @@ impl Node<Option<cst::Policy>> {
-
-        // now case 1: if `policy` is `Some(Err(e))`, there is a slot where there shouldn't be;
-        // report that as `UnexpectedTemplate`
-        match policy {
-            Some(Err(ast::UnexpectedSlotError::FoundSlot(slot))) => {
-                errs.push(ToASTError::new(
+    pub fn to_policy(&self, id: ast::PolicyID) -> Result<ast::StaticPolicy, ParseErrors> {
+        let maybe_template = self.to_policy_template(id);
+        let maybe_policy = maybe_template.map(ast::StaticPolicy::try_from);
+        match maybe_policy {
+            // Successfully parsed a static policy
+            Ok(Ok(p)) => Ok(p),
+            // The source parsed as a template, but not a static policy
+            Ok(Err(ast::UnexpectedSlotError::FoundSlot(slot))) => {
+                Err(ParseErrors::singleton(ToASTError::new(
                     ToASTErrorKind::UnexpectedTemplate {
                         slot: slot.id.into(),
                     },
                     slot.loc.unwrap_or_else(|| self.loc.clone()),
-                ));
-                None
+                )))
+            }
+            // The source failed to parse completely. If the parse errors include
+            // `SlotsInConditionClause` also add an `UnexpectedTemplate` error.
+            Err(mut errs) => {
+                let new_errs = errs
+                    .iter()
+                    .filter_map(|err| match err {
+                        ParseError::ToAST(err) => match err.kind() {
+                            ToASTErrorKind::SlotsInConditionClause { slot, .. } => {
+                                Some(ToASTError::new(
+                                    ToASTErrorKind::UnexpectedTemplate { slot: slot.clone() },
+                                    err.source_loc().clone(),
+                                ))
+                            }
+                            _ => None,
+                        },
+                        _ => None,
+                    })
+                    .collect::<Vec<_>>();
+                errs.extend(new_errs);
+                Err(errs)
             }
-            // in other cases, we're done reporting errors, so we can return the policy, if we have one
-            Some(Ok(p)) => Some(p),
-            None => None,
         }
     }
 
     /// Convert `cst::Policy` to `ast::Template`. Works for inline policies as
     /// well, which will become templates with 0 slots
-    pub fn to_policy_template(
-        &self,
-        id: ast::PolicyID,
-        errs: &mut ParseErrors,
-    ) -> Option<ast::Template> {
-        // if `self` doesn't have data, nothing we can do here, just propagate
-        // the `None`; we don't need to signal an error, because one was already
-        // signaled when the `Node` without data was created
-        let policy = self.as_inner()?;
+    pub fn to_policy_template(&self, id: ast::PolicyID) -> Result<ast::Template, ParseErrors> {
+        let policy = self.ok_or_missing()?;
```

Comment:
Same comment as above, we didn't use to throw `MissingNodeData` here ever, now we do, is that desirable?